### ISSUE TITLE
Updating gcc 7 to gcc 8 with homebrew breaks coordinates module

### ISSUE THREAD

**Issue by johncoxon** [spacepy/spacepy#41]
I get an error when trying to use a `Coords` object in Mojave, having just updated. Running the following two blocks of code:
```
from spacepy import coordinates
coordinates.Coords(np.random.random((100,3)), np.random.random(100), 'car')
```

Gives the following traceback:
```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-8-f437b8d7fe1e> in <module>()
----> 1 coordinates.Coords(np.random.random((100,3)), np.random.random(100), 'car')

/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/coordinates.py in __init__(self, data, dtype, carsph, units, ticks)
     75     def __init__(self, data, dtype, carsph, units=None, ticks=None):
     76 
---> 77         from . import irbempy as op
     78         from spacepy.irbempy import SYSAXES_TYPES as typedict
     79 

/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/irbempy/__init__.py in <module>()
     69 
     70 
---> 71 from .irbempy import *
     72 

/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/irbempy/irbempy.py in <module>()
     26 import spacepy.datamodel as dm
     27 try:
---> 28     from . import irbempylib as oplib
     29 except ImportError:
     30     if 'sphinx' in sys.argv[0]:

ImportError: dlopen(/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/irbempy/irbempylib.cpython-36m-darwin.so, 2): Library not loaded: /usr/local/opt/gcc/lib/gcc/7/libgfortran.4.dylib
  Referenced from: /anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/irbempy/irbempylib.cpython-36m-darwin.so
  Reason: image not found
```

I tried `brew install gcc` as this looks like a gfortran issue to me, and I got:
```
(python3) home-one:~ john$ brew install gcc
Warning: gcc 8.2.0 is already installed and up-to-date
To reinstall 8.2.0, run `brew reinstall gcc`
```

I tried `pip uninstall spacepy` and `pip install spacepy` and that did not help either.

Does anyone have any ideas?

**Comment by johncoxon** [comment]
(I've edited the above to have a minimum reproducible example, rather than the original comment, which took the traceback directly from the code I was using it in – sorry about that!)

**Comment by johncoxon** [comment]
One thing that does occur to me is that I did update most of my homebrew packages when I updated to Mojave, so is it possible that irbempy just needs to be updated to work with gcc 8.2.0? I note that `/usr/local/opt/gcc/lib/gcc/8/libgfortran.5.dylib` exists on my machine whereas `/usr/local/opt/gcc/lib/gcc/7/libgfortran.4.dylib` does not!

**Comment by jtniehof** [comment]
Thanks for the details. Just to clarify, the time sequence was: 1) At some point in the past you installed SpacePy (using pip?) 2) Then you upgraded to Mojave 3) Then the previously-installed SpacePy didn't work, and you tried a reinstall?

The irbem library (which we use for coordinate conversions) is linked against the Fortran runtime library at build time, and the upgrade swapped versions of the runtime library. So to have it break on the upgrade isn't necessarily surprising. What's more surprising is that a reinstall didn't fix things. When you say "pip install spacepy did not help" does that mean the pip install looks like it went successfully, but you're still getting the error, or the pip install didn't work?

I'm seeing two obvious possibilities here. Most likely, either pip somehow reused a build directory from the previous install and just reinstalled the previously built library without rebuilding it, or the pip install didn't replace the SpacePy you had installed. The latter is easier to check...try pip uninstall spacepy again, and then without an install, try to import SpacePy. If it succeeds, there's an old version lurking around.

I'm going to open a ticket to make the installer behave better in the case of the old build directory hanging around...may not be relevant, but probably a good idea.

@balarsen , do you have thoughts as the Mac dude?

**Comment by johncoxon** [comment]
So I will say that when I ran `pip uninstall spacepy` and then ran `pip install spacepy`, I didn't get anything print to the Terminal except for:
```
Collecting spacepy
Requirement already satisfied: networkx in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.1)
Requirement already satisfied: ffnet in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (0.8.3)
Requirement already satisfied: numpy>=1.4 in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (1.14.3)
Requirement already satisfied: matplotlib>=0.99 in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.2.2)
Requirement already satisfied: scipy>=0.7 in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (1.1.0)
Requirement already satisfied: python-dateutil in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.7.3)
Requirement already satisfied: h5py in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.7.1)
Requirement already satisfied: decorator>=4.1.0 in /anaconda2/envs/python3/lib/python3.6/site-packages (from networkx->spacepy) (4.3.0)
Requirement already satisfied: cycler>=0.10 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (0.10.0)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (2.2.0)
Requirement already satisfied: pytz in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (2018.4)
Requirement already satisfied: six>=1.10 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (1.11.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (1.0.1)
Requirement already satisfied: setuptools in /anaconda2/envs/python3/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib>=0.99->spacepy) (39.1.0)
distributed 1.21.8 requires msgpack, which is not installed.
Installing collected packages: spacepy
Successfully installed spacepy-0.1.6
You are using pip version 10.0.1, however version 18.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

I thought it was odd that I hadn't gotten any messages about fortran compiling things, so we might be on the right lines here?

**Comment by johncoxon** [comment]
So I tried doing `pip uninstall spacepy` followed by `pip install --no-cache-dir spacepy` and got the following (still no fortran outputs, but is clearly now downloading the relevant package):
```
Collecting spacepy
  Downloading https://files.pythonhosted.org/packages/f7/10/4a9a43014d25c1ce08aad136da2515eb120a26105dec5e8f42b5986c22a3/spacepy-0.1.6.tar.gz (11.1MB)
    100% |████████████████████████████████| 11.1MB 49.4MB/s 
Requirement already satisfied: numpy>=1.4 in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (1.14.3)
Requirement already satisfied: scipy>=0.7 in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (1.1.0)
Requirement already satisfied: matplotlib>=0.99 in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.2.2)
Requirement already satisfied: h5py in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.7.1)
Requirement already satisfied: ffnet in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (0.8.3)
Requirement already satisfied: networkx in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.1)
Requirement already satisfied: python_dateutil in /anaconda2/envs/python3/lib/python3.6/site-packages (from spacepy) (2.7.3)
Requirement already satisfied: cycler>=0.10 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (0.10.0)
Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (2.2.0)
Requirement already satisfied: pytz in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (2018.4)
Requirement already satisfied: six>=1.10 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (1.11.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /anaconda2/envs/python3/lib/python3.6/site-packages (from matplotlib>=0.99->spacepy) (1.0.1)
Requirement already satisfied: decorator>=4.1.0 in /anaconda2/envs/python3/lib/python3.6/site-packages (from networkx->spacepy) (4.3.0)
Requirement already satisfied: setuptools in /anaconda2/envs/python3/lib/python3.6/site-packages (from kiwisolver>=1.0.1->matplotlib>=0.99->spacepy) (39.1.0)
distributed 1.21.8 requires msgpack, which is not installed.
Installing collected packages: spacepy
  Running setup.py install for spacepy ... done
Successfully installed spacepy-0.1.6
You are using pip version 10.0.1, however version 18.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

Afterwards, I got a different error when running my sample code:

```
In [1]: from spacepy import coordinates

In [2]: coordinates.Coords(np.random.random((100,3)), np.random.random(100), 'car')
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-f437b8d7fe1e> in <module>()
----> 1 coordinates.Coords(np.random.random((100,3)), np.random.random(100), 'car')

/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/coordinates.py in __init__(self, data, dtype, carsph, units, ticks)
     75     def __init__(self, data, dtype, carsph, units=None, ticks=None):
     76 
---> 77         from . import irbempy as op
     78         from spacepy.irbempy import SYSAXES_TYPES as typedict
     79 

/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/irbempy/__init__.py in <module>()
     69 
     70 
---> 71 from .irbempy import *
     72 

/anaconda2/envs/python3/lib/python3.6/site-packages/spacepy/irbempy/irbempy.py in <module>()
     26 import spacepy.datamodel as dm
     27 try:
---> 28     from . import irbempylib as oplib
     29 except ImportError:
     30     if 'sphinx' in sys.argv[0]:

ImportError: cannot import name 'irbempylib'
```

**Comment by johncoxon** [comment]
Sorry, and to clarify, the time sequence was:
0) I installed gcc with `brew` at some point in the past.
1) I installed SpacePy using `pip` at some point in the past.
2) I upgraded to Mojave
3) I updated gcc to gcc 8 using `brew`
4) SpacePy did not work and I came here

**Comment by balarsen** [comment]
Mojave is still not approved for me at work so can't directly test but my suspicion is that other stuff is broken as well. As the brew/macports gets all kind of messed up on update in the past. Did you reinstall xcode command line utilities? `xcode-select -p`

Things that I would try:
- can you compile C programs under the new setup?
- clone the spacepy source repo
- python setup.py install --user    (capture the output)

**Comment by johncoxon** [comment]
`xcode-select -p` returns `/Library/Developer/CommandLineTools`

I never compile C programs, so I'm not sure. I'm not sure what 'clone the spacepy source repo' means, and `python setup.py install --user` returns `python: can't open file 'setup.py': [Errno 2] No such file or directory`.

My gut feeling that this is still probably because I updated `gcc` from 7.x to 8.x and now `/usr/local/opt/gcc/lib/gcc/8/libgfortran.5.dylib` exists on my machine whereas `/usr/local/opt/gcc/lib/gcc/7/libgfortran.4.dylib` – I think if I could get spacepy to actually recompile the fortran bits when I ran `pip install spacepy` it'd probably fix the problem, but I can't work out how to do that (the `--no-cache-dir` attempt above didn't result in any fortran compiler messages outputting to the Terminal).

**Comment by drsteve** [comment]
Just a quick translation of @balarsen's last comment.

By "clone the repo" he means getting the source code from the git repository here on github. There are a couple of ways of doing that. One would be to use git to "clone" (copy) the full repository. Assuming you have git installed on your system the syntax for doing that would be:
```git clone https://github.com/spacepy/spacepy.git```
That will grab the source code, and the full history of the repository, and put it in a subdirectory called ```spacepy```. So if you choose to do that then you'll need to make sure you're in an appropriate directory when you enter that command.

An alternative, that's easier if you're not used to developing or working with git, is to go to [https://github.com/spacepy/spacepy](https://github.com/spacepy/spacepy) and click the green button that says "clone or download". A box will drop down that has info for cloning, and under that it'll say "Download ZIP". Click that, download the ZIP file and then unzip into a directory.

Either of these methods should give you a directory called ```spacepy``` that contains all of the source code. One of the files in there will be ```setup.py```, so if you change to that directory and then type:
```python setup.py install --user```
it should attempt the full build, including the Fortran code. If you can try that and let us know what happens (maybe redirect the output into a file, since there should be a lot of output).

An easier way of starting might be to just try pip again, but use
```pip install spacepy --user --ignore-installed```
That way pip shouldn't check what's already installed and might try to compile the fortran. My first guess would be that pip is seeing no change in the fortran source and so isn't rebuilding, even though it needs to. Another way of forcing it to rebuild would be to track down the locations where pip has installed spacepy and to remove any spacepy-related directories or files, then try ```pip install spacepy``` again. That *should* attempt to compile the fortran.

**Comment by balarsen** [comment]
maybe related issue: https://discourse.pymc.io/t/problems-with-pymc3-after-updating-to-macos-mojave/1994

**Comment by johncoxon** [comment]
I rolled back to a Time Machine backup from just after I installed Mojave but before I updated gcc from 7 to 8, and that solved the issue – so it's the gcc update that was the problem, not the Mojave installation.

I've renamed the issue to better describe the issue I had!

**Comment by drsteve** [comment]
Closing, as this issue has gone stale.
Other issues had raised the fact that the fortran had some comments in a style that was no longer sorted by default in gcc 8, but this was fixed in #106.

Basically, the underlying issues with gcc 8 should be fixed in master currently and we'll be pushing out a new release to pip soon, so this shouldn't happen again.

If anyone thinks this should be reopened, just let me know here.
